src/hotspot/share/c1/c1_LIRGenerator.cpp

```diff
@@ -1345,11 +1345,12 @@ void LIRGenerator::do_getObjectSize(Intrinsic* x) {
 
   // Instance case: the layout helper gives us instance size almost directly,
   // but we need to mask out the _lh_instance_slow_path_bit.
-  __ convert(Bytecodes::_i2l, layout, result_reg);
 
   assert((int) Klass::_lh_instance_slow_path_bit < BytesPerLong, "clear bit");
-  jlong mask = ~(jlong) right_n_bits(LogBytesPerLong);
-  __ logical_and(result_reg, LIR_OprFact::longConst(mask), result_reg);
+
+  LIR_Opr mask = load_immediate(~(jint) right_n_bits(LogBytesPerLong), T_INT);
```

Comment:
Just to note: taken a look at generated code to check why this helps with wrong result issue. It appears jdk suffers from the problem of inlining of this code into caller method since this lines are never executed for small array case

Reply:
I don't understand what you mean here :)

Reply:
Well, OK, never mind. I'm just saying this change workarounds some generic c1 issue